`Job.close()` / `Job.wait()` silently swallow an external `CancelledError` (bpo-42130 family)

**Affected:** `Job._close` ; Reproduced on **aiojobs 1.4.0** (latest release, commit `8844625`) with Python 3.12+.

### What happened?

`Job._close` cancels the inner task and awaits it inside a bare `except asyncio.CancelledError: pass`:

```python
# aiojobs/_job.py
self._task.cancel()
try:
    async with asyncio_timeout(timeout):
        await self._task
except asyncio.CancelledError:
    pass
```

If the task that is calling `await job.close()` (or `await job.wait()`, which routes failures through `_close`) is itself **cancelled from the outside** while suspended at `await self._task`, that external `CancelledError` is caught and discarded. `close()` returns normally and the caller keeps running — although it was cancelled.

This is the same class of defect as CPython's [bpo-42130](https://github.com/python/cpython/issues/86296) (`asyncio.wait_for` swallowing a cancellation): the `except` cannot distinguish *"the inner task finished by cancellation"* (expected) from *"I, the closer, was cancelled externally"* (must propagate). The practical consequence is that a `TaskGroup` teardown, an `asyncio.timeout`, or any cancel scope wrapping `job.close()` is silently defeated.

###  Reproduction (plain asyncio, no extra deps)

```python
import asyncio
from aiojobs._scheduler import Scheduler

async def main():
    sched = Scheduler()
    started = asyncio.Event()

    async def coro():
        started.set()
        try:
            await asyncio.sleep(1000)
        except asyncio.CancelledError:
            await asyncio.sleep(0)   # inner needs one extra step to finish,
            raise                    # keeping the closer parked at `await self._task`

    job = await sched.spawn(coro())
    await started.wait()

    async def closer():
        await job.close()            # -> _close: self._task.cancel(); await self._task

    t = asyncio.create_task(closer())
    await asyncio.sleep(0)           # closer enters _close, cancels inner, parks at await
    await asyncio.sleep(0)
    delivered = t.cancel()           # external cancellation lands on the parked closer
    try:
        await t
        outcome = "returned normally"
    except asyncio.CancelledError:
        outcome = "raised CancelledError"

    print(f"cancel delivered={delivered}  outcome={outcome}  t.cancelled()={t.cancelled()}")
    assert t.cancelled(), "BUG: external cancellation of the closer was swallowed"
    await sched.close()

asyncio.run(main())
```

Output:

```
cancel delivered=True  outcome=returned normally  t.cancelled()=False
AssertionError: BUG: external cancellation of the closer was swallowed
```

### How it was found, and that it is schedule-dependent

Found with [simloom](https://pypi.org/project/simloom/), a deterministic asyncio scheduler. Exploring 120 seeds, the external cancel lands on the still-suspended closer in 8 of them — and in **every one of those the cancellation is swallowed (0 of 8 propagate)**. So the bug fires whenever its (timing-specific) precondition is met; the plain-asyncio repro above pins that timing manually. The failing schedule replays byte-for-byte from its seed.

### Suggested fix

Distinguish an external cancellation of the closer from the inner task's expected cancellation — the bpo-42130 remedy — and re-raise in the former case:

```python
try:
    async with asyncio_timeout(timeout):
        await self._task
except asyncio.CancelledError:
    if asyncio.current_task().cancelling() > 0:   # the closer itself was cancelled
        raise
    # otherwise: the inner task's cancellation completed — expected
```

I verified the discrimination: with this guard, the same 120-seed exploration drops from 8 swallowing schedules to **0**, while the normal close path is unaffected.

### Relationship to existing issues

The bare `except asyncio.CancelledError: pass` has been present since 2017 and is unchanged at HEAD; the 2025 refactor (PR #546) restructured around it but left these lines intact. The closest existing report, #252, is the *inverse* concern (`close()` *propagating* a cancellation); its reporter's workaround is in fact `try: await scheduler.close() except asyncio.CancelledError: pass` — i.e. the very pattern flagged here. I found no existing report of external caller-cancellation being swallowed, and nothing in `CHANGES.rst`.

--- a/CHANGES/606.bugfix
+++ b/CHANGES/606.bugfix
@@ -1,1 +1,1 @@
-Fixed ``Job.close()`` and ``Job.wait()`` swallowing an external cancellation of the task calling them; awaiting the job's task no longer conflates the job's own cancellation with a cancellation aimed at the caller.
+Fixed ``Job.close()`` and ``Job.wait()`` swallowing an external cancellation of the calling task (exact on Python 3.11+ via ``Task.cancelling()``; on older versions the cancellation propagates whenever the job is still running).

--- a/CHANGES/606.bugfix
+++ b/CHANGES/606.bugfix
@@ -1,0 +1,1 @@
+Fixed ``Job.close()`` and ``Job.wait()`` swallowing an external cancellation of the task calling them; awaiting the job's task no longer conflates the job's own cancellation with a cancellation aimed at the caller.

--- a/aiojobs/_job.py
+++ b/aiojobs/_job.py
@@ -130,25 +130,29 @@ class Job(Generic[_T]):
         self._task.cancel()
         # self._scheduler is None after _done_callback()
         scheduler = self._scheduler
-        # Wait for the task through an intermediate future instead of
-        # awaiting it directly: awaiting the task would deliver
-        # CancelledError both when the cancelled task finishes and when
-        # the current task is cancelled from the outside (cancelling a
-        # task that is awaiting another task also cancels the awaited
-        # task, so the task state does not tell the two apart).  The
-        # waiter is only ever cancelled by a cancellation aimed at the
-        # current task, which must keep propagating so that callers of
-        # close()/wait() stay cancellable.
-        waiter: "asyncio.Future[None]" = asyncio.get_running_loop().create_future()
-
-        def _on_completion(task: "asyncio.Task[_T]") -> None:
-            if not waiter.cancelled():
-                waiter.set_result(None)
-
-        self._task.add_done_callback(_on_completion)
         try:
             async with asyncio_timeout(timeout):
-                await waiter
+                if sys.version_info >= (3, 11):
+                    await self._task
+                else:
+                    # Cancelling the current task would be forwarded to
+                    # self._task through _fut_waiter, making the two
+                    # cancellation sources indistinguishable; shield the
+                    # job so an external cancellation leaves it running
+                    # and detectable below.
+                    await asyncio.shield(self._task)
+        except asyncio.CancelledError:
+            # Either the cancelled job finished or the task running
+            # close() was itself cancelled; re-raise in the second case
+            # so callers stay cancellable.
+            if sys.version_info >= (3, 11):
+                ctask = asyncio.current_task()
+                if ctask is not None and ctask.cancelling():
+                    raise
+            elif not self._task.done():
+                # The job is still running, so the CancelledError cannot
+                # have come from awaiting it.
+                raise
         except asyncio.TimeoutError as exc:
             if self._explicit:
                 raise
@@ -163,14 +167,9 @@ class Job(Generic[_T]):
             # there's no timeout. self._scheduler will now be None though.
             assert scheduler is not None
             scheduler.call_exception_handler(context)
-            return
-        finally:
-            self._task.remove_done_callback(_on_completion)
-        if self._task.cancelled():
-            return
-        exc2 = self._task.exception()
-        if exc2 is not None and (self._explicit or not isinstance(exc2, Exception)):
-            raise exc2
+        except Exception:
+            if self._explicit:
+                raise
 
     def _start(self) -> None:
         assert self._task is None

--- a/aiojobs/_job.py
+++ b/aiojobs/_job.py
@@ -130,11 +130,25 @@ class Job(Generic[_T]):
         self._task.cancel()
         # self._scheduler is None after _done_callback()
         scheduler = self._scheduler
+        # Wait for the task through an intermediate future instead of
+        # awaiting it directly: awaiting the task would deliver
+        # CancelledError both when the cancelled task finishes and when
+        # the current task is cancelled from the outside (cancelling a
+        # task that is awaiting another task also cancels the awaited
+        # task, so the task state does not tell the two apart).  The
+        # waiter is only ever cancelled by a cancellation aimed at the
+        # current task, which must keep propagating so that callers of
+        # close()/wait() stay cancellable.
+        waiter: "asyncio.Future[None]" = asyncio.get_running_loop().create_future()
+
+        def _on_completion(task: "asyncio.Task[_T]") -> None:
+            if not waiter.cancelled():
+                waiter.set_result(None)
+
+        self._task.add_done_callback(_on_completion)
         try:
             async with asyncio_timeout(timeout):
-                await self._task
-        except asyncio.CancelledError:
-            pass
+                await waiter
         except asyncio.TimeoutError as exc:
             if self._explicit:
                 raise
@@ -149,9 +163,14 @@ class Job(Generic[_T]):
             # there's no timeout. self._scheduler will now be None though.
             assert scheduler is not None
             scheduler.call_exception_handler(context)
-        except Exception:
-            if self._explicit:
-                raise
+            return
+        finally:
+            self._task.remove_done_callback(_on_completion)
+        if self._task.cancelled():
+            return
+        exc2 = self._task.exception()
+        if exc2 is not None and (self._explicit or not isinstance(exc2, Exception)):
+            raise exc2
 
     def _start(self) -> None:
         assert self._task is None

--- a/aiojobs/_job.py
+++ b/aiojobs/_job.py
@@ -147,7 +147,8 @@ class Job(Generic[_T]):
             # so callers stay cancellable.
             if sys.version_info >= (3, 11):
                 ctask = asyncio.current_task()
-                if ctask is not None and ctask.cancelling():
+                assert ctask is not None
+                if ctask.cancelling() > 0:
                     raise
             elif not self._task.done():
                 # The job is still running, so the CancelledError cannot

--- a/tests/test_job.py
+++ b/tests/test_job.py
@@ -263,8 +263,7 @@ async def test_job_close_cancelled_from_outside(scheduler: Scheduler) -> None:
     # the job task, which is blocked until unblock is set.
     await cancel_seen.wait()
     closer.cancel()
-    with pytest.raises(asyncio.CancelledError):
-        await closer
+    await asyncio.wait({closer})
     assert closer.cancelled()
 
     unblock.set()
@@ -297,8 +296,7 @@ async def test_job_close_cancelled_from_outside_completion_race(
     # up and detaches them.
     unblock.set()
     closer.cancel()
-    with pytest.raises(asyncio.CancelledError):
-        await closer
+    await asyncio.wait({closer})
     assert closer.cancelled()
     with suppress(asyncio.CancelledError):
         await job.wait()

--- a/tests/test_job.py
+++ b/tests/test_job.py
@@ -1,4 +1,5 @@
 import asyncio
+import sys
 from collections.abc import Awaitable
 from contextlib import suppress
 from typing import Callable, NoReturn
@@ -271,12 +272,16 @@ async def test_job_close_cancelled_from_outside(scheduler: Scheduler) -> None:
         await job.wait()
 
 
+@pytest.mark.skipif(
+    sys.version_info < (3, 11),
+    reason="distinguishing this race needs Task.cancelling()",
+)
 async def test_job_close_cancelled_from_outside_completion_race(
     scheduler: Scheduler,
 ) -> None:
     """The job task may finish in the same loop iteration the caller of
-    close() is cancelled in; the already scheduled completion callback
-    then finds the waiter cancelled and must leave it alone."""
+    close() is cancelled in; the external cancellation must still
+    propagate."""
     cancel_seen = asyncio.Event()
     unblock = asyncio.Event()
 

--- a/tests/test_job.py
+++ b/tests/test_job.py
@@ -239,6 +239,39 @@ async def test_job_close_closed(make_scheduler: _MakeScheduler) -> None:
     await job.close()
 
 
+async def test_job_close_cancelled_from_outside(scheduler: Scheduler) -> None:
+    """An external cancellation of the task running close() must propagate.
+
+    The CancelledError raised by awaiting the job's cancelled task is
+    expected and swallowed, but a cancellation aimed at the caller of
+    close() itself is not ours to swallow.
+    """
+    cancel_seen = asyncio.Event()
+    unblock = asyncio.Event()
+
+    async def coro() -> None:
+        try:
+            await asyncio.sleep(60)
+        except asyncio.CancelledError:
+            cancel_seen.set()
+            await unblock.wait()
+            raise
+
+    job = await scheduler.spawn(coro())
+    closer = asyncio.ensure_future(job.close())
+    # Once the job saw the cancellation, close() is suspended awaiting
+    # the job task, which is blocked until unblock is set.
+    await cancel_seen.wait()
+    closer.cancel()
+    with pytest.raises(asyncio.CancelledError):
+        await closer
+    assert closer.cancelled()
+
+    unblock.set()
+    with suppress(asyncio.CancelledError):
+        await job.wait()
+
+
 async def test_job_await_closed(scheduler: Scheduler) -> None:
     async def coro() -> int:
         return 5

--- a/tests/test_job.py
+++ b/tests/test_job.py
@@ -272,6 +272,65 @@ async def test_job_close_cancelled_from_outside(scheduler: Scheduler) -> None:
         await job.wait()
 
 
+async def test_job_close_cancelled_from_outside_completion_race(
+    scheduler: Scheduler,
+) -> None:
+    """The job task may finish in the same loop iteration the caller of
+    close() is cancelled in; the already scheduled completion callback
+    then finds the waiter cancelled and must leave it alone."""
+    cancel_seen = asyncio.Event()
+    unblock = asyncio.Event()
+
+    async def coro() -> None:
+        try:
+            await asyncio.sleep(60)
+        except asyncio.CancelledError:
+            cancel_seen.set()
+            await unblock.wait()
+            raise
+
+    job = await scheduler.spawn(coro())
+    closer = asyncio.ensure_future(job.close())
+    await cancel_seen.wait()
+    # Unblock the job before cancelling the closer: the job task then
+    # completes and schedules its done callbacks before the closer wakes
+    # up and detaches them.
+    unblock.set()
+    closer.cancel()
+    with pytest.raises(asyncio.CancelledError):
+        await closer
+    assert closer.cancelled()
+    with suppress(asyncio.CancelledError):
+        await job.wait()
+
+
+async def test_job_close_timeout_source_traceback(
+    make_scheduler: _MakeScheduler,
+) -> None:
+    """In debug mode the close timeout report carries the source traceback."""
+    loop = asyncio.get_running_loop()
+    loop.set_debug(True)
+    try:
+        handler = mock.Mock()
+        scheduler = await make_scheduler(close_timeout=0.01, exception_handler=handler)
+
+        async def coro() -> None:
+            try:
+                await asyncio.sleep(60)
+            except asyncio.CancelledError:
+                await asyncio.sleep(60)
+
+        job = await scheduler.spawn(coro())
+        await scheduler.close()
+        assert job.closed
+        assert handler.called
+        context = handler.call_args[0][1]
+        assert context["message"] == "Job closing timed out"
+        assert "source_traceback" in context
+    finally:
+        loop.set_debug(False)
+
+
 async def test_job_await_closed(scheduler: Scheduler) -> None:
     async def coro() -> int:
         return 5

